### PR TITLE
feat: use use-query-params for state management 

### PR DESCRIPTION
**Pull request by mediremi**
Closes https://jira.dhis2.org/browse/DHIS2-10444

We currently store filtering and pagination state for the sent and received messages pages in both query parameters and component state, which adds complexity by forcing us to sync multiple sources of truth. The aim of this PR is to simplify state handling by using query parameters (through the [`use-query-params` package](https://github.com/pbeshai/use-query-params)) as the sole source of truth.

### CODE REVIEW

File: src/views/received_sms_list/ReceivedSmsList.js [hunk 1]
```diff
@@ -52,7 +52,20 @@ const query = {
...
+        setQueryParams({ phoneNumber, page: 1 })
+    }
+    const setStatus = status => {
+        setQueryParams({ status, page: 1 })
+    }
+    const handleFilterReset = () => {
+        setQueryParams({
+            phoneNumber: undefined,
+            status: undefined,
+        })
```
Comment (Mohammer5) [inline src/views/received_sms_list/ReceivedSmsList.js:67]: I think this also set the page to `1`

**Comment by ismay** [comment]
I don't yet have the time for an elaborate review (hopefully tomorrow), but from my perspective, the core goal here is to reduce the internal filtering state to a single source of truth. So instead of having the component's state, and the query params state, the goal here is to just use the query params for storing state. Reason for doing is to reduce the complexity we noticed in dealing with / synchronizing the multiple sources of truth.

That does mean that we need to immediately fetch according to the query params on mount, and do a debounced fetch at least for updates to the phone number input (so that it doesn't fetch for every keypress there).

@mediremi Would you agree with the above perspective? Anything I've missed or misrepresented?

File: src/views/received_sms_list/ReceivedSmsList.js (hunk 1 above)
Comment (mediremi) [reply src/views/received_sms_list/ReceivedSmsList.js:67]: Good catch :+1:

**Comment by mediremi** [comment]
@ismay that's exactly right :ok_hand: and I've updated the PR description to make that clearer. We also had the issue of various components [having to know how routing is done in order to update their state](https://github.com/dhis2/sms-configuration-app/blob/8340ce90049b4c40603713359cb2d513cd762ddb/src/views/sent_sms_list/StatusFilter.js#L25-L32), which has been fixed by moving the updating of query param state to the [`ReceivedSmsList`](https://github.com/dhis2/sms-configuration-app/blob/8f4155c62444ba964e3631d07b41a622a8db1fa4/src/views/received_sms_list/ReceivedSmsList.js#L57-L62) and [`SentSmsList`](https://github.com/dhis2/sms-configuration-app/blob/8f4155c62444ba964e3631d07b41a622a8db1fa4/src/views/sent_sms_list/SentSmsList.js#L41-L43) components.

**Review by ismay — CHANGES_REQUESTED** [review]
Sorry for the late review (again 😅). Changes look great, nice work. I think it's reduced a lot of the complexity. One case that I think we still have to deal with is a double refetch in the ReceivedSmsList:

- Say you've updated the statusfilter & the phonenumber
- Now press the reset button
- A refetch will be done twice (as status has changed and a refetch immediately occurs, and after 500ms the delayed phonenumber refetch will be done as well)

Not sure what would be the best way to deal with that. Also, I think it'd be nice to have a couple of explanatory comments on the useEffect usage in the ReceivedSmsList, so that future devs can know at a glance what's going on there.

That's everything that stood out to me. After addressing the above I'd say it's good to merge! What do you think @Mohammer5?

**Comment by mediremi** [comment]
[I've combined the two `useEffect`s into one using `lodash.debounce` with `leading: true`.](https://github.com/dhis2/sms-configuration-app/pull/217/files#diff-0908bb4466971618a26633a474f27ec2d98e3945a056246a3ec046e631abe92eR78-R85) Using `leading: true` means that the initial fetch is not delayed and so other filter params are not impacted.

So that's both issues (double refetch & non-obvious use of `useEffect`) resolved :sunglasses:

**Comment by ismay** [comment]
Nice. One thing that I think could improve this even further is using https://github.com/xnimorz/use-debounce. I think their `useDebouncedCallback` would be a bit more descriptive, making the code easier to comprehend for an outsider. Plus, it's smaller than `lodash.debounce`, well tested, maintained and widely used.

- https://bundlephobia.com/result?p=use-debounce@5.2.1 (useDebouncedCallback, 893B)
- https://bundlephobia.com/result?p=lodash.debounce@4.0.8 (938B)

Not that that amount of bytes really matters, but use-debounce seems better suited to our usecase as well, so it seems like an overall win to me. What do you think?

**Comment by mediremi** [comment]
You're right that the `use-debounce` API is more intuitive than the `useCallback` + `lodash.debounce` combo - for example it's not immediately obvious that `useCallback(debounce(fn))` is incorrect (note the lack of `[]` for the deps param) and `use-debounce` helps us avoid such footguns. The smaller size is an added bonus too :ok_hand:

**Review by ismay — APPROVED** [review]
Perfect, looks great!

**Pull request merged**